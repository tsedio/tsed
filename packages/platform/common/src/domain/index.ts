export * from "./AnyToPromiseWithCtx";
export * from "./PlatformContext";
export * from "./PlatformRouteDetails";
export * from "./ControllerProvider";
export * from "./EndpointMetadata";
export * from "./HandlerMetadata";
export * from "./PropertyMetadata";
