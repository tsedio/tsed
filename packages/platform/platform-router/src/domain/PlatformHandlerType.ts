export enum PlatformHandlerType {
  CUSTOM = "custom",
  ENDPOINT = "endpoint",
  MIDDLEWARE = "middleware",
  ERR_MIDDLEWARE = "err:middleware",
  CTX_FN = "context",
  RESPONSE_FN = "response",
  RAW_FN = "raw:middleware",
  RAW_ERR_FN = "raw:err:middleware"
}
