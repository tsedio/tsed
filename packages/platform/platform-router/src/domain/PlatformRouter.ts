import {isString} from "@tsed/core";
import {Inject, Injectable, InjectorService, type Provider, ProviderScope, Scope} from "@tsed/di";
import {concatPath} from "@tsed/schema";

import {formatMethod} from "../utils/formatMethod.js";
import {PlatformHandlerMetadata} from "./PlatformHandlerMetadata.js";
import type {PlatformLayerOptions} from "./PlatformLayer.js";
import {PlatformLayer} from "./PlatformLayer.js";
import type {SinglePathType} from "./SinglePathType.js";

function printHandler(handler: any) {
  return handler.toString().split("{")[0].trim();
}

@Injectable()
@Scope(ProviderScope.INSTANCE)
export class PlatformRouter {
  readonly layers: PlatformLayer[] = [];
  provider: Provider;
  #isBuilt = false;

  constructor(@Inject(InjectorService) protected readonly injector: InjectorService) {}

  use(...handlers: any[]) {
    const layer = handlers.reduce(
      (layer: PlatformLayer, item) => {
        if (isString(item) || item instanceof RegExp) {
          layer.path = item;
          layer.basePath = item;
        } else {
          if (item instanceof PlatformRouter) {
            layer.router = item;
            layer.setLayers(item.layers);

            if (!this.provider && item.provider) {
              layer.path = concatPath(layer.path, item.provider.path);
            } else {
              layer.path = layer.path || item.provider.path;
            }
          } else {
            item = PlatformHandlerMetadata.from(this.injector, item);
          }

          layer.handlers.push(item);
        }

        return layer;
      },
      new PlatformLayer({
        method: "use",
        basePath: this.provider?.path || "/",
        provider: this.provider
      })
    );

    this.layers.push(layer);

    return this;
  }

  addRoute(method: string, path: SinglePathType, handlers: any[], opts: PlatformLayerOptions = {}) {
    const layer = new PlatformLayer({
      provider: this.provider,
      method: formatMethod(method),
      path,
      handlers: handlers.map((input) => {
        return PlatformHandlerMetadata.from(this.injector, input, opts);
      }),
      opts
    });

    this.layers.push(layer);

    return this;
  }

  all(path: SinglePathType, ...handlers: any[]) {
    return this.addRoute("all", path, handlers);
  }

  get(path: SinglePathType, ...handlers: any[]) {
    return this.addRoute("get", path, handlers);
  }

  post(path: SinglePathType, ...handlers: any[]) {
    return this.addRoute("post", path, handlers);
  }

  put(path: SinglePathType, ...handlers: any[]) {
    return this.addRoute("put", path, handlers);
  }

  delete(path: SinglePathType, ...handlers: any[]) {
    return this.addRoute("delete", path, handlers);
  }

  patch(path: SinglePathType, ...handlers: any[]) {
    return this.addRoute("patch", path, handlers);
  }

  head(path: SinglePathType, ...handlers: any[]) {
    return this.addRoute("head", path, handlers);
  }

  options(path: SinglePathType, ...handlers: any[]) {
    return this.addRoute("options", path, handlers);
  }

  statics(path: string, options: any) {
    return this.addRoute("statics", path, [], options);
  }

  inspect() {
    return this.layers.map((layer) => {
      const obj = layer.inspect();

      return {
        ...obj,
        handlers: obj.handlers.map(printHandler),
        path: concatPath(this.provider?.path, obj.path)
      };
    });
  }

  isBuilt() {
    if (this.#isBuilt) {
      return true;
    }

    this.#isBuilt = true;

    return false;
  }
}
