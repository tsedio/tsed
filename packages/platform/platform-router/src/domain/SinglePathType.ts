export type SinglePathType = string | RegExp;
