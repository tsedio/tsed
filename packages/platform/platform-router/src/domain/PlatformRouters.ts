import {getValue, Hooks, Type} from "@tsed/core";
import {ControllerProvider, GlobalProviders, Injectable, InjectorService, Provider, ProviderType, TokenProvider} from "@tsed/di";
import {concatPath, getOperationsRoutes} from "@tsed/schema";
import {useContextHandler} from "../utils/useContextHandler";
import {PlatformLayer} from "./PlatformLayer";
import {PlatformRouter} from "./PlatformRouter";

let AUTO_INC = 0;

function getInjectableRouter(injector: InjectorService, provider: Provider): PlatformRouter {
  return injector.get(provider.tokenRouter)!;
}

function createTokenRouter(provider: ControllerProvider) {
  return (provider.tokenRouter = provider.tokenRouter || `${provider.name}_ROUTER_${AUTO_INC++}`);
}

function createInjectableRouter(injector: InjectorService, provider: ControllerProvider): PlatformRouter {
  const tokenRouter = createTokenRouter(provider);

  if (injector.has(tokenRouter)) {
    return getInjectableRouter(injector, provider);
  }

  const router = injector.invoke<PlatformRouter>(PlatformRouter);
  router.provider = provider;

  return injector
    .add(tokenRouter, {
      useValue: router
    })
    .invoke<PlatformRouter>(tokenRouter);
}

GlobalProviders.createRegistry(ProviderType.CONTROLLER, ControllerProvider, {
  onInvoke(provider: ControllerProvider, locals: any, {injector}) {
    const router = createInjectableRouter(injector, provider);
    locals.set(PlatformRouter, router);
  }
});

@Injectable()
export class PlatformRouters {
  readonly hooks = new Hooks();

  constructor(protected readonly injector: InjectorService) {}

  prebuild() {
    this.injector.getProviders(ProviderType.CONTROLLER).forEach((provider: ControllerProvider) => {
      createInjectableRouter(this.injector, provider);
    });
  }

  from(token: TokenProvider, parentMiddlewares: any[] = []) {
    const {injector} = this;
    const provider = injector.getProvider<ControllerProvider>(token)!;

    if (!provider) {
      throw new Error("Token not found in the provider registry");
    }

    const router = createInjectableRouter(injector, provider);

    if (router.isBuilt()) {
      return router;
    }

    const useBefore = getValue(provider, "middlewares.useBefore", []);

    const {children} = provider;

    getOperationsRoutes(provider.token).forEach((operationRoute) => {
      const {endpoint} = operationRoute;
      const {beforeMiddlewares, middlewares: mldwrs, afterMiddlewares} = endpoint;

      const useBefore = getValue(provider, "middlewares.useBefore", []);
      const use = getValue(provider, "middlewares.use", []);
      const useAfter = getValue(provider, "middlewares.useAfter", []);

      const handlers = this.hooks.alter(
        "alterEndpointHandlers",
        [
          ...parentMiddlewares,
          ...useBefore,
          ...beforeMiddlewares,
          ...use,
          ...mldwrs,
          operationRoute.endpoint,
          ...afterMiddlewares,
          ...useAfter
        ],
        [operationRoute],
        this
      );

      router.addRoute(
        operationRoute.method,
        operationRoute.path || "",
        [
          useContextHandler(($ctx) => {
            $ctx.endpoint = operationRoute.endpoint;
          }),
          ...handlers
        ],
        operationRoute
      );
    });

    const middlewares: any[] = [...parentMiddlewares, ...useBefore];

    children.forEach((token: Type<any>) => {
      const nested = this.from(token, middlewares);

      router.use(nested);
    });

    return router;
  }

  getLayers(router: PlatformRouter): PlatformLayer[] {
    return router.layers
      .flatMap((layer) => {
        if (layer.router) {
          return this.getLayers(layer.router).map((subLayer: PlatformLayer) => {
            return new PlatformLayer({
              ...subLayer,
              path: concatPath(layer.path, subLayer.path)
            });
          });
        }

        return new PlatformLayer(layer);
      })
      .map((layer) => {
        const handlers = layer.handlers.map((handlerMetadata) => {
          // set path on handler metadata to retrieve it later in $ctx
          handlerMetadata.path = layer.path;

          return this.hooks.alter("alterHandler", handlerMetadata);
        });

        layer.set(handlers);

        return layer;
      });
  }
}
