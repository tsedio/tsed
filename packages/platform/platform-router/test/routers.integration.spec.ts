import {PlatformContext, PlatformHandlerMetadata, PlatformTest, useResponseHandler} from "@tsed/common";
import {catchError} from "@tsed/core";
import {Controller, InjectorService} from "@tsed/di";
import {UseBefore} from "@tsed/platform-middlewares";
import {Context, PlatformParams} from "@tsed/platform-params";
import {Delete, Get, Head, Options, Patch, Post, Put} from "@tsed/schema";
import {PlatformRouter} from "../src/domain/PlatformRouter";
import {AlterEndpointHandlersArg, PlatformRouters} from "../src/domain/PlatformRouters";

@Controller("/nested")
class NestedController {
  @Get("/")
  get() {}

  @Post("/")
  post() {}

  @Put("/:id")
  put() {}

  @Delete("/:id")
  delete() {}

  @Head("/:id")
  head() {}

  @Options("/:id")
  option() {}

  @Patch("/:id")
  patch() {}
}

@Controller({path: "/controller", children: [NestedController]})
@UseBefore(function useBefore() {})
class MyController {
  @Get("/")
  get(@Context() $ctx: Context) {
    return $ctx;
  }

  @Post("/")
  post() {}

  @Put("/:id")
  put() {}

  @Delete("/:id")
  delete() {}

  @Head("/:id")
  head() {}

  @Options("/:id")
  option() {}

  @Patch("/:id")
  patch() {}
}

function createAppRouterFixture() {
  const injector = new InjectorService();
  const platformRouters = injector.invoke<PlatformRouters>(PlatformRouters);
  const platformParams = injector.invoke<PlatformParams>(PlatformParams);
  const appRouter = injector.invoke<PlatformRouter>(PlatformRouter);

  injector.addProvider(NestedController, {});

  platformRouters.hooks.on("alterEndpointHandlers", (handlers: AlterEndpointHandlersArg) => {
    handlers.after.push(useResponseHandler(() => "hello"));
    return handlers;
  });
  platformRouters.hooks.on("alterHandler", (handlerMetadata: PlatformHandlerMetadata) => {
    if (handlerMetadata.isInjectable()) {
      return platformParams.compileHandler(handlerMetadata);
    }

    return handlerMetadata.handler;
  });

  return {injector, appRouter, platformRouters, platformParams};
}

describe("routers integration", () => {
  beforeEach(() => PlatformTest.create());
  afterEach(() => PlatformTest.reset());
  describe("getLayers()", () => {
    it("should declare router", () => {
      const {injector, platformRouters} = createAppRouterFixture();
      injector.addProvider(MyController, {});

      const hookStub = jest.fn().mockImplementation((o) => o);

      platformRouters.hooks.on("alterEndpointHandlers", hookStub);

      const router = platformRouters.from(MyController);

      expect(hookStub).toHaveBeenCalled();
      expect(router.inspect()).toMatchSnapshot();
    });
    it("should declare router - appRouter", async () => {
      const {injector, appRouter, platformRouters} = createAppRouterFixture();
      injector.addProvider(MyController, {});

      const router = platformRouters.from(MyController);

      appRouter.use("/rest", router);

      // prebuild controllers
      platformRouters.prebuild();

      // returns layers
      const layers = platformRouters.getLayers(appRouter);

      expect(layers.length).toEqual(14);
      expect(layers.map((l) => l.inspect())).toMatchSnapshot();
      expect(layers.find((layer) => layer.method == "use")).toEqual(undefined);

      const args = layers[0].getArgs();
      expect(layers[0].isProvider()).toEqual(true);
      expect(args[0]).toEqual("/rest/controller");
      expect(layers[0].getArgs().length).toEqual(5);

      const $ctx = PlatformTest.createRequestContext();
      const result = await (args[3] as any)!({$ctx});

      expect(result).toBeInstanceOf(PlatformContext);
    });
    it("should throw an error when the controller isn't found", () => {
      const {platformRouters} = createAppRouterFixture();

      const error: any = catchError(() => platformRouters.from(class Test {}));

      expect(error?.message).toEqual("Token not found in the provider registry");
    });
  });

  describe("use()", () => {
    it("should call method", () => {
      const injector = new InjectorService();
      injector.addProvider(NestedController, {});

      const router = new PlatformRouter(injector);

      router.use("/hello", function h() {});

      expect(router.inspect()).toEqual([
        {
          basePath: "/hello",
          handlers: ["h"],
          method: "use",
          opts: {},
          path: "/hello"
        }
      ]);
    });
  });
  describe("get()", () => {
    it("should call method", () => {
      const {appRouter} = createAppRouterFixture();

      appRouter.get("/hello", function h() {});

      expect(appRouter.inspect()).toEqual([
        {
          basePath: "",
          handlers: ["h"],
          method: "get",
          opts: {},
          path: "/hello"
        }
      ]);
    });
  });
  describe("post()", () => {
    it("should call method", () => {
      const {appRouter} = createAppRouterFixture();

      appRouter.post("/hello", function h() {});

      expect(appRouter.inspect()).toEqual([
        {
          basePath: "",
          handlers: ["h"],
          method: "post",
          opts: {},
          path: "/hello"
        }
      ]);
    });
  });
  describe("put()", () => {
    it("should call method", () => {
      const {appRouter} = createAppRouterFixture();

      appRouter.put("/hello", function h() {});

      expect(appRouter.inspect()).toEqual([
        {
          basePath: "",
          handlers: ["h"],
          method: "put",
          opts: {},
          path: "/hello"
        }
      ]);
    });
  });
  describe("patch()", () => {
    it("should call method", () => {
      const {appRouter} = createAppRouterFixture();

      appRouter.patch("/hello", function h() {});

      expect(appRouter.inspect()).toEqual([
        {
          basePath: "",
          handlers: ["h"],
          method: "patch",
          opts: {},
          path: "/hello"
        }
      ]);
    });
  });
  describe("head()", () => {
    it("should call method", () => {
      const {appRouter} = createAppRouterFixture();

      appRouter.head("/hello", function h() {});

      expect(appRouter.inspect()).toEqual([
        {
          basePath: "",
          handlers: ["h"],
          method: "head",
          opts: {},
          path: "/hello"
        }
      ]);
    });
  });
  describe("delete()", () => {
    it("should call method", () => {
      const {appRouter} = createAppRouterFixture();

      appRouter.delete("/hello", function h() {});

      expect(appRouter.inspect()).toEqual([
        {
          basePath: "",
          handlers: ["h"],
          method: "delete",
          opts: {},
          path: "/hello"
        }
      ]);
    });
  });
  describe("option()", () => {
    it("should call method", () => {
      const {appRouter} = createAppRouterFixture();

      appRouter.options("/hello", function h() {});

      expect(appRouter.inspect()).toEqual([
        {
          basePath: "",
          handlers: ["h"],
          method: "options",
          opts: {},
          path: "/hello"
        }
      ]);
    });
  });
  describe("all()", () => {
    it("should call method", () => {
      const {appRouter} = createAppRouterFixture();

      appRouter.all("/hello", function h() {});

      expect(appRouter.inspect()).toEqual([
        {
          basePath: "",
          handlers: ["h"],
          method: "all",
          opts: {},
          path: "/hello"
        }
      ]);
    });
  });
  describe("statics()", () => {
    it("should call method", () => {
      const {appRouter} = createAppRouterFixture();

      appRouter.statics("/hello", {root: "root"});

      expect(appRouter.inspect()).toEqual([
        {
          basePath: "",
          handlers: [],
          method: "statics",
          opts: {
            root: "root"
          },
          path: "/hello"
        }
      ]);
    });
  });
});
