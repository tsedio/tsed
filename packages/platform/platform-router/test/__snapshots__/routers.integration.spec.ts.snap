// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`routers integration > getLayers() > should declare router - appRouter 1`] = `
[
  {
    "basePath": "/rest",
    "handlers": [
      "",
      "useBefore",
      "MyController.get",
      "",
    ],
    "method": "get",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:get",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller",
  },
  {
    "basePath": "/rest",
    "handlers": [
      "",
      "useBefore",
      "MyController.post",
      "",
    ],
    "method": "post",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:post",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller",
  },
  {
    "basePath": "/rest",
    "handlers": [
      "",
      "useBefore",
      "MyController.put",
      "",
    ],
    "method": "put",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:put",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller/:id",
  },
  {
    "basePath": "/rest",
    "handlers": [
      "",
      "useBefore",
      "MyController.delete",
      "",
    ],
    "method": "delete",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:delete",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller/:id",
  },
  {
    "basePath": "/rest",
    "handlers": [
      "",
      "useBefore",
      "MyController.head",
      "",
    ],
    "method": "head",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:head",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller/:id",
  },
  {
    "basePath": "/rest",
    "handlers": [
      "",
      "useBefore",
      "MyController.option",
      "",
    ],
    "method": "options",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:option",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller/:id",
  },
  {
    "basePath": "/rest",
    "handlers": [
      "",
      "useBefore",
      "MyController.patch",
      "",
    ],
    "method": "patch",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:patch",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller/:id",
  },
  {
    "basePath": "/rest/controller",
    "handlers": [
      "",
      "useBefore",
      "NestedController.get",
      "",
    ],
    "method": "get",
    "opts": {
      "basePath": "/nested",
      "endpoint": "NestedController:get",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested",
  },
  {
    "basePath": "/rest/controller",
    "handlers": [
      "",
      "useBefore",
      "NestedController.post",
      "",
    ],
    "method": "post",
    "opts": {
      "basePath": "/nested",
      "endpoint": "NestedController:post",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested",
  },
  {
    "basePath": "/rest/controller",
    "handlers": [
      "",
      "useBefore",
      "NestedController.put",
      "",
    ],
    "method": "put",
    "opts": {
      "basePath": "/nested",
      "endpoint": "NestedController:put",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested/:id",
  },
  {
    "basePath": "/rest/controller",
    "handlers": [
      "",
      "useBefore",
      "NestedController.delete",
      "",
    ],
    "method": "delete",
    "opts": {
      "basePath": "/nested",
      "endpoint": "NestedController:delete",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested/:id",
  },
  {
    "basePath": "/rest/controller",
    "handlers": [
      "",
      "useBefore",
      "NestedController.head",
      "",
    ],
    "method": "head",
    "opts": {
      "basePath": "/nested",
      "endpoint": "NestedController:head",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested/:id",
  },
  {
    "basePath": "/rest/controller",
    "handlers": [
      "",
      "useBefore",
      "NestedController.option",
      "",
    ],
    "method": "options",
    "opts": {
      "basePath": "/nested",
      "endpoint": "NestedController:option",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested/:id",
  },
  {
    "basePath": "/rest/controller",
    "handlers": [
      "",
      "useBefore",
      "NestedController.patch",
      "",
    ],
    "method": "patch",
    "opts": {
      "basePath": "/nested",
      "endpoint": "NestedController:patch",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested/:id",
  },
]
`;

exports[`routers integration > getLayers() > should declare router 1`] = `
[
  {
    "basePath": "",
    "handlers": [
      "",
      "useBefore",
      "MyController.get",
      "",
    ],
    "method": "get",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:get",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller",
  },
  {
    "basePath": "",
    "handlers": [
      "",
      "useBefore",
      "MyController.post",
      "",
    ],
    "method": "post",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:post",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller",
  },
  {
    "basePath": "",
    "handlers": [
      "",
      "useBefore",
      "MyController.put",
      "",
    ],
    "method": "put",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:put",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller/:id",
  },
  {
    "basePath": "",
    "handlers": [
      "",
      "useBefore",
      "MyController.delete",
      "",
    ],
    "method": "delete",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:delete",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller/:id",
  },
  {
    "basePath": "",
    "handlers": [
      "",
      "useBefore",
      "MyController.head",
      "",
    ],
    "method": "head",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:head",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller/:id",
  },
  {
    "basePath": "",
    "handlers": [
      "",
      "useBefore",
      "MyController.option",
      "",
    ],
    "method": "options",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:option",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller/:id",
  },
  {
    "basePath": "",
    "handlers": [
      "",
      "useBefore",
      "MyController.patch",
      "",
    ],
    "method": "patch",
    "opts": {
      "basePath": "/controller",
      "endpoint": "MyController:patch",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller/:id",
  },
  {
    "basePath": "/controller",
    "handlers": [
      "[object Object]",
    ],
    "method": "use",
    "opts": {},
    "path": "/controller/nested",
  },
]
`;

exports[`routers integration getLayers() should declare router - appRouter 1`] = `
Array [
  Object {
    "basePath": "/rest",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.get",
      "",
    ],
    "method": "get",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:get",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller",
  },
  Object {
    "basePath": "/rest",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.post",
      "",
    ],
    "method": "post",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:post",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller",
  },
  Object {
    "basePath": "/rest",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.put",
      "",
    ],
    "method": "put",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:put",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller/:id",
  },
  Object {
    "basePath": "/rest",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.delete",
      "",
    ],
    "method": "delete",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:delete",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller/:id",
  },
  Object {
    "basePath": "/rest",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.head",
      "",
    ],
    "method": "head",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:head",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller/:id",
  },
  Object {
    "basePath": "/rest",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.option",
      "",
    ],
    "method": "options",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:option",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller/:id",
  },
  Object {
    "basePath": "/rest",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.patch",
      "",
    ],
    "method": "patch",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:patch",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/rest/controller/:id",
  },
  Object {
    "basePath": "/rest/controller",
    "handlers": Array [
      "",
      "useBefore",
      "NestedController.get",
      "",
    ],
    "method": "get",
    "opts": Object {
      "basePath": "/nested",
      "endpoint": "NestedController:get",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested",
  },
  Object {
    "basePath": "/rest/controller",
    "handlers": Array [
      "",
      "useBefore",
      "NestedController.post",
      "",
    ],
    "method": "post",
    "opts": Object {
      "basePath": "/nested",
      "endpoint": "NestedController:post",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested",
  },
  Object {
    "basePath": "/rest/controller",
    "handlers": Array [
      "",
      "useBefore",
      "NestedController.put",
      "",
    ],
    "method": "put",
    "opts": Object {
      "basePath": "/nested",
      "endpoint": "NestedController:put",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested/:id",
  },
  Object {
    "basePath": "/rest/controller",
    "handlers": Array [
      "",
      "useBefore",
      "NestedController.delete",
      "",
    ],
    "method": "delete",
    "opts": Object {
      "basePath": "/nested",
      "endpoint": "NestedController:delete",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested/:id",
  },
  Object {
    "basePath": "/rest/controller",
    "handlers": Array [
      "",
      "useBefore",
      "NestedController.head",
      "",
    ],
    "method": "head",
    "opts": Object {
      "basePath": "/nested",
      "endpoint": "NestedController:head",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested/:id",
  },
  Object {
    "basePath": "/rest/controller",
    "handlers": Array [
      "",
      "useBefore",
      "NestedController.option",
      "",
    ],
    "method": "options",
    "opts": Object {
      "basePath": "/nested",
      "endpoint": "NestedController:option",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested/:id",
  },
  Object {
    "basePath": "/rest/controller",
    "handlers": Array [
      "",
      "useBefore",
      "NestedController.patch",
      "",
    ],
    "method": "patch",
    "opts": Object {
      "basePath": "/nested",
      "endpoint": "NestedController:patch",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "NestedController",
    },
    "path": "/rest/controller/nested/:id",
  },
]
`;

exports[`routers integration getLayers() should declare router 1`] = `
Array [
  Object {
    "basePath": "",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.get",
      "",
    ],
    "method": "get",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:get",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller",
  },
  Object {
    "basePath": "",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.post",
      "",
    ],
    "method": "post",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:post",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller",
  },
  Object {
    "basePath": "",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.put",
      "",
    ],
    "method": "put",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:put",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller/:id",
  },
  Object {
    "basePath": "",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.delete",
      "",
    ],
    "method": "delete",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:delete",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller/:id",
  },
  Object {
    "basePath": "",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.head",
      "",
    ],
    "method": "head",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:head",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller/:id",
  },
  Object {
    "basePath": "",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.option",
      "",
    ],
    "method": "options",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:option",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller/:id",
  },
  Object {
    "basePath": "",
    "handlers": Array [
      "",
      "useBefore",
      "MyController.patch",
      "",
    ],
    "method": "patch",
    "opts": Object {
      "basePath": "/controller",
      "endpoint": "MyController:patch",
      "operationPath": "[object Object]",
      "paramsTypes": "[object Object]",
      "token": "MyController",
    },
    "path": "/controller/:id",
  },
  Object {
    "basePath": "/controller",
    "handlers": Array [
      "[object Object]",
    ],
    "method": "use",
    "opts": Object {},
    "path": "/controller/nested",
  },
]
`;
